viz/MLSMapVisualization.cpp

```diff
@@ -123,27 +126,85 @@ struct DataHold : public MLSMapVisualization::Data
+
+                        float height = patch->getMax();
+                        float nheight = neighborpatch->getMax();
+
+                        //geode.setColor( horizontalCellColor );
+                        Eigen::Vector3f pos (xypos.x(), xypos.y(), height);
+                        Eigen::Vector3f posnormal = patch->getNormal();
+                        Eigen::Vector3f npos (nxypos.x(), nxypos.y(), nheight);
+                        Eigen::Vector3f nposnormal = neighborpatch->getNormal();
+
+                        // adding the neighbot first to make the SmoothingVisitor work properly
+                        geode.addVertex( osg::Vec3f(pos.x(),npos.y(),npos.z()), osg::Vec3f(nposnormal.x(),nposnormal.y(),nposnormal.z()) );
+                        geode.addVertex( osg::Vec3f(pos.x(),pos.y(),pos.z()), osg::Vec3f(posnormal.x(),posnormal.y(),posnormal.z()) );
+
+                        lastheight = height;
+
+
+                    }else{
+                        geode.closeTriangleStrip();
+                    }
+                }
+                geode.closeTriangleStrip();
+            }
+            //compute normals
+            osgUtil::SmoothingVisitor smoothingVisitor;
+            smoothingVisitor.apply(geode);
+        }
+
+
+
+    void visualize(vizkit3d::PatchesGeode& geode, int levelOffset) const
+    {
+        Vector2ui num_cell = mls.getNumCells();
+
+            //const GridMap<SPListST> &mls = *this;
+            
+            for (size_t x = 0; x < num_cell.x(); x++)
             {
-                typedef typename MLSMap<Type>::CellType Cell;
-                const Cell &list = mls.at(x, y);
-
-                Vector2d pos(0.00, 0.00);
-                // Calculate the position of the cell center.
-                pos = (maps::grid::Index(x, y).cast<double>() + maps::grid::Vector2d(0.5, 0.5)).array() * mls.getResolution().array();
-                geode.setPosition(pos.x(), pos.y());
-                for (typename Cell::const_iterator it = list.begin(); it != list.end(); it++)
+                for (size_t y = 0; y < num_cell.y(); y++)
                 {
-                    PatchVisualizer::visualize(geode, *it);
-                } // for(SPList ...)
-            } // for(y ...)
-        } // for(x ...)        
+                    typedef typename MLSMap<Type>::CellType Cell;
+                    const Cell &list = mls.at(x, y);
+
+                    Vector2d pos(0.00, 0.00);
+                    // Calculate the position of the cell center.
+                    pos = (maps::grid::Index(x, y).cast<double>() + maps::grid::Vector2d(0.5, 0.5)).array() * mls.getResolution().array();
+                    geode.setPosition(pos.x(), pos.y());
+                    if (list.size()){
+                        for (typename Cell::const_iterator it = list.begin()+levelOffset; it != list.end(); it++)
```

Comment:
I fail to get this part. It drops the lowest patches on the MLS, currently just the first patch in case connected surface is true. What is the purpose here?